### PR TITLE
Honor dictionary key policy when serializing

### PR DESCRIPTION
**Pull request by layomia**
Fixes https://github.com/dotnet/corefx/issues/38840.

cc @ahsonkhan @pranavkm

### CODE REVIEW

File: src/System.Text.Json/tests/Serialization/DictionaryTests.KeyPolicy.cs [hunk 1]
```diff
@@ -0,0 +1,103 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using Xunit;
+
+namespace System.Text.Json.Serialization.Tests
+{
+    public static class DictionaryKeyPolicyTests
```
Comment (pranavkm) [inline src/System.Text.Json/tests/Serialization/DictionaryTests.KeyPolicy.cs]: Could you also add a test to verify this works with `JsonExtensionData` dictionary property?

Comment (pranavkm) [reply src/System.Text.Json/tests/Serialization/DictionaryTests.KeyPolicy.cs]: And one more to see what the behavior is when you have case insensitive keys + camel case. Just to ensure that the keys are unaffected by the case insentivity

**Review by pranavkm — COMMENTED** [review]
:shipit:

File: src/System.Text.Json/tests/Serialization/DictionaryTests.KeyPolicy.cs [hunk 2]
```diff
@@ -99,5 +101,38 @@ public static void NullNamePolicy()
             Assert.Throws<InvalidOperationException>(() => JsonSerializer.Deserialize<Dictionary<string, int>>(@"{""onlyKey"": 1}", options));
             Assert.Throws<InvalidOperationException>(() => JsonSerializer.Serialize(new Dictionary<string, int> { { "onlyKey", 1 } }, options));
         }
+
+        [Fact]
+        public static void KeyConflictDeserialize_LastValueWins()
```
Comment (pranavkm) [inline src/System.Text.Json/tests/Serialization/DictionaryTests.KeyPolicy.cs]: @JamesNK  could you have a look to see if these scenarios look good? @layomia realized we're missing a feature equivalent for https://www.newtonsoft.com/json/help/html/P_Newtonsoft_Json_Serialization_NamingStrategy_ProcessExtensionDataNames.htm, but that could come in the future

**Review by pranavkm — COMMENTED** [review]
Looks good to me

File: src/System.Text.Json/tests/Serialization/DictionaryTests.KeyPolicy.cs (hunk 1 above)
Comment (layomia) [reply src/System.Text.Json/tests/Serialization/DictionaryTests.KeyPolicy.cs]: > Could you also add a test to verify this works with `JsonExtensionData` dictionary property?

As is, this won't work by default with `JsonExtensionData` dictionary properties. This is consistent with Json.NET. With Json.NET, one needs to specify that naming strategies should apply to extension data: https://www.newtonsoft.com/json/help/html/P_Newtonsoft_Json_Serialization_NamingStrategy_ProcessDictionaryKeys.htm.

To support this with our options, we might need to add a `ExtensionDataPropertyNamePolicy` or something similar.

Comment (layomia) [reply src/System.Text.Json/tests/Serialization/DictionaryTests.KeyPolicy.cs]: > And one more to see what the behavior is when you have case insensitive keys + camel case. Just to ensure that the keys are unaffected by the case insentivity

This behavior is covered well in the [PropertyNameTests](https://github.com/dotnet/corefx/blob/master/src/System.Text.Json/tests/Serialization/PropertyNameTests.cs).

On a related note, this is the current behavior for dictionary key policies that lead to two or more keys having the same output key (e.g. both myInt and MyInt => myInt):
- On serialization, we write all keys and their values
i.e `new Dictionary<string, int> { { "myInt", 1 }, { "MyInt", 2 } }` => `@"{""myInt"": 1,""myInt"": 2}"`.
This is consistent with Json.NET (at least by default)
- On deserialization, the last value wins
i.e parsing `@"{""myInt"": 1,""MyInt"": 2}"` => `new Dictionary<string, int> { { "myInt", 2 } }`

Comment (layomia) [reply src/System.Text.Json/tests/Serialization/DictionaryTests.KeyPolicy.cs]: cc @steveharter @ahsonkhan

Comment (steveharter) [reply src/System.Text.Json/tests/Serialization/DictionaryTests.KeyPolicy.cs]: > To support this with our options, we might need to add a ExtensionDataPropertyNamePolicy or something similar.

This was previously discussed - the extension data should keep its original naming. But yes, we could always add another naming policy in the future.

Comment (steveharter) [reply src/System.Text.Json/tests/Serialization/DictionaryTests.KeyPolicy.cs]: @layomia I also have a [PR ](https://github.com/dotnet/corefx/pull/39042) out that deals with extension data. We may have to modify that once your change goes in to ignore the dictionary key policy.

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandlePropertyName.cs [hunk 3]
```diff
@@ -45,6 +45,11 @@ public static partial class JsonSerializer
                     if (options.DictionaryKeyPolicy != null)
                     {
                         keyName = options.DictionaryKeyPolicy.ConvertName(keyName);
```
Comment (JamesNK) [inline src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandlePropertyName.cs]: Wait a second, why is the dictionary name policy used when deserializing? When was that decided?

**Comment by layomia** [comment]
> Wait a second, why is the dictionary name policy used when deserializing? When was that decided?

Since it's a trivial change to not use the dictionary key policy when deserializing, I'm merging this to fix https://github.com/dotnet/corefx/issues/38840 while this is decided on.
cc @steveharter, @pranavkm, @ahsonkhan, @rynowak, @JamesNK

**Comment by layomia** [comment]
> > Since it's a trivial change to not use the dictionary key policy when deserializing, I'm merging this to fix #38840 while this is decided on.
> 
> Did we ever close on this?

We closed on it here: https://github.com/dotnet/corefx/issues/39207

**Pull request merged**